### PR TITLE
fix(core): include Flickable-clipped items in tab focus traversal

### LINKED ISSUE #10321: Tab navigation skips FocusScope elements outside visible viewport in Flickable/ScrollView
## Description

When using keyboard Tab navigation in a `Flickable` or `ScrollView`, `FocusScope` elements that are outside the visible viewport are skipped entirely. The focus jumps only between visible elements, making it impossible to navigate to off-screen focusable content via keyboard.

## Expected Behavior

Tab navigation should cycle through all focusable elements in document order, automatically scrolling the viewport to bring the newly focused element into view (similar to browser behavior with focusable elements in scrollable containers).

Steps to Reproduce

1. Run the example above
2. Click on "Button 1" to give it focus
3. Press Tab repeatedly to navigate through buttons
4. Observe that Tab only cycles through buttons visible in the viewport
5. Buttons below the fold (Button 5-10) are never reached via Tab

## Actual Behavior

- Tab navigation only visits `FocusScope` elements within the visible viewport
- Off-screen focusable elements are skipped entirely
- No automatic scrolling occurs to bring focused elements into view

## Workarounds Attempted (None Successful)

1. **`focus-changed-event` callback**: Never fires for off-screen elements since they are skipped
2. **Custom Tab interception at parent level**: Slint's internal focus management runs before user code
3. **Programmatic `focus()` calls**: Element must be visible first for focus to work

## Suggested Solution

When Tab navigation moves focus to the next `FocusScope`, automatically scroll the containing `Flickable`/`ScrollView` to ensure the focused element is visible. This could be similar to browser behavior where focusing an element via Tab automatically scrolls it into view.

### Reproducible Code (if applicable)

```slint
import { VerticalBox } from "std-widgets.slint";

component FocusableButton inherits Rectangle {
    in property <string> label;
    
    width: 200px;
    height: 50px;
    background: focus-scope.has-focus ? #4f39f6 : #e5e5e5;
    border-radius: 8px;
    
    focus-scope := FocusScope {
        width: 100%;
        height: 100%;
        
        key-pressed(event) => {
            if event.text == Key.Return {
                debug("Button pressed:", root.label);
                return accept;
            }
            reject
        }
    }
    
    Text {
        text: root.label;
        color: focus-scope.has-focus ? white : black;
    }
}

export component App inherits Window {
    width: 400px;
    height: 300px;
    
    Flickable {
        width: 100%;
        height: 100%;
        viewport-height: 800px;
        
        VerticalBox {
            padding: 20px;
            spacing: 20px;
            
            FocusableButton { label: "Button 1"; }
            FocusableButton { label: "Button 2"; }
            FocusableButton { label: "Button 3"; }
            FocusableButton { label: "Button 4"; }
            FocusableButton { label: "Button 5"; }
            FocusableButton { label: "Button 6"; }
            FocusableButton { label: "Button 7"; }
            FocusableButton { label: "Button 8"; }
            FocusableButton { label: "Button 9"; }
            FocusableButton { label: "Button 10"; }
        }
    }
}
```

### Environment Details

- Slint Version: 1.9.2
- Platform/OS:  macOS / Linux (NixOS)
- Programming Language: Rust
- Backend/Renderer: Skia OpenGL

### Product Impact

This significantly affects accessibility for keyboard-only users. Any scrollable content with focusable elements becomes partially inaccessible via keyboard navigation.

### PR DESCRIPTION
**Pull request by amirHdev**
## Summary

Fixes #10321.

Tab navigation in `Window::move_focus()` filtered candidates using `ItemRc::is_visible()`.  
For `FocusScope`s inside a `Flickable`, off-screen items are clipped and therefore not visible, so they were skipped before focus could be assigned. This prevented keyboard users from reaching items below the fold.

This change:
- Adds `ItemRc::is_visible_or_clipped_by_flickable()`
- Uses it for `FocusReason::TabNavigation` in `Window::move_focus()`
- Keeps existing behavior for other focus reasons:
  - `Programmatic`: always allowed
  - others: still require `is_visible()`

Once focus is accepted, existing `try_scroll_into_visible()` logic can reveal the newly focused item.

## Tests

Added regression test:

- `tests/cases/issues/issue_10321_tab_focus_flickable.slint`

The test verifies tab traversal visits all focus scopes inside a `Flickable` in order, including items initially outside the viewport (no skipping).

## Validation

Executed:

- `RUST_BACKTRACE=1 cargo test -p test-driver-rust -- issue_10321_tab_focus_flickable`

### CODE REVIEW

File: tests/cases/issues/issue_10321_tab_focus_flickable.slint [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+            width: parent.width;
+            height: 60px;
+            focus-gained(_) => { root.focused-index = 4; }
+        }
+        FocusScope {
+            y: 350px;
+            width: parent.width;
+            height: 60px;
+            focus-gained(_) => { root.focused-index = 5; }
+        }
```
Comment (LeonMatthes) [inline tests/cases/issues/issue_10321_tab_focus_flickable.slint]: This should really be in a VerticalLayout with a for loop

File: tests/cases/issues/issue_10321_tab_focus_flickable.slint [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+        width: 100%;
+        height: 100%;
+        viewport-height: 420px;
+
+        FocusScope {
+            y: 0px;
+            width: parent.width;
+            height: 60px;
+            focus-gained(_) => { root.focused-index = 0; }
+        }
```
Comment (LeonMatthes) [inline tests/cases/issues/issue_10321_tab_focus_flickable.slint]:
```suggestion
        FocusScope {
            y: 0px;
            width: parent.width;
            height: 60px;
            focus-gained(_) => { root.focused-index = 0; }
            Rectangle { background: "red"; }
        }
```

It would be nice to have differently colored rectangles in these FocusScopes so that this test can be debugged in the live-preview or slint-viewer.

File: tests/cases/issues/issue_10321_tab_focus_flickable.slint [hunk 3]
```diff
@@ -0,0 +1,75 @@
+// Copyright © SixtyFPS GmbH <info@slint.dev>
+// SPDX-License-Identifier: GPL-3.0-only OR LicenseRef-Slint-Royalty-free-2.0 OR LicenseRef-Slint-Software-3.0
+
+export component TestCase inherits Window {
```
Comment (LeonMatthes) [inline tests/cases/issues/issue_10321_tab_focus_flickable.slint:6]: This test behaves unexpectedly when opened in the live-preview.
When pressing tab it only cycles between the first two elements in the flickable, but then focuses an item in the live-previews button bar.

It does work correctly in slint-viewer, so this could be an existing issue in our tab handling.

Comment (LeonMatthes) [reply tests/cases/issues/issue_10321_tab_focus_flickable.slint:6]: @amirHdev I don't think you have to necessarily fix this in this PR, but it would be great if you could give it a look.

File: internal/core/item_tree.rs [hunk 4]
```diff
@@ -369,6 +369,46 @@ impl ItemRc {
...
+        while let Some(item_rc) = parent {
+            parent = item_rc.parent_item(ParentItemTraversalMode::StopAtPopups);
+            ancestors.push(item_rc);
+        }
+
+        for ancestor in ancestors.into_iter().rev() {
+            let clips_children = ancestor.borrow().as_ref().clips_children();
+            if !clips_children {
+                continue;
+            }
```
Comment (LeonMatthes) [inline internal/core/item_tree.rs:393]: nit: this could be moved into the first loop to avoid looping over irrelevant ancestors twice

File: internal/core/item_tree.rs [hunk 5]
```diff
@@ -369,6 +369,46 @@ impl ItemRc {
...
+            clip = ancestor_geometry.intersection(&clip).unwrap_or_default();
+            let clip = clip.to_box2d();
+
+            let is_visible = !clip.is_empty()
+                && clip.max.x >= geometry.min.x
+                && clip.max.y >= geometry.min.y
+                && clip.min.x <= geometry.max.x
+                && clip.min.y <= geometry.max.y;
+            if !is_visible {
+                return ancestor.downcast::<crate::items::Flickable>().is_some();
```
Comment (LeonMatthes) [inline internal/core/item_tree.rs:405]: If my understanding is correct, this loop goes outside-in from the window towards the item and stops at the first item that reduces the clip by so much that the item is no longer visible.

However, we then stop traversing there and don't check if anything else will reduce the clip even further, which may cause the item to be invisible after all. That could be a problematic edge case.
But I think that would probably be very tricky to solve fully, and is probably rather rare, so the current approach should be good-enough for now.

**Review by LeonMatthes — COMMENTED** [review]
Hi @amirHdev , thank you for the contribution!

This is going in a good direction, I think the test could be refactored a bit more, and this may not solve all edge cases correctly, but I think it should solve the problem in the majority of cases 👍

**Comment by amirHdev** [comment]
@LeonMatthes Thanks for your feedback.
consider to review it please.

Thanks

File: internal/core/item_tree.rs [hunk 6]
```diff
@@ -369,6 +369,49 @@ impl ItemRc {
...
+
+            let is_visible = !clip.is_empty()
+                && clip.max.x >= geometry.min.x
+                && clip.max.y >= geometry.min.y
+                && clip.min.x <= geometry.max.x
+                && clip.min.y <= geometry.max.y;
+            if !is_visible {
+                if !is_flickable {
+                    return false;
+                }
```
Comment (LeonMatthes) [inline internal/core/item_tree.rs]: This heuristic unfortunately is a little bit too simple.
Because it uses the global clip rect, instead of local coordinates, if there is any item between the flickable and the target that has clip enabled, this will immediately return false, because as soon as the clip makes the item invisible, the is_visible check will never return true again, thereby running into this branch.

You can test this by modifying the test case to add a rectangle around the focusscope that has clipping enabled:
```slint
            for _[index] in [0, 1, 2, 3, 4, 5] : Rectangle {
                width: 100%;
                height: 60px;
                clip: true;
                FocusScope {
                    width: 100%;
                    height: 60px;
                    focus-gained(_) => { root.focused-index = index; }

                    Rectangle {
                        width: 100%;
                        height: 100%;
                        background: index == 0 ? red
                            : index == 1 ? orange
                            : index == 2 ? yellow
                            : index == 3 ? green
                            : index == 4 ? blue
                            : purple;
                    }
                }
            }
```

The previous approach is probably good enough so you can stick with what you used before.

File: tests/cases/issues/issue_10321_tab_focus_flickable.slint [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+slint_testing::send_keyboard_string_sequence(&instance, "\t");
+assert_eq!(instance.get_focused_index(), 2);
+slint_testing::send_keyboard_string_sequence(&instance, "\t");
+assert_eq!(instance.get_focused_index(), 3);
+slint_testing::send_keyboard_string_sequence(&instance, "\t");
+assert_eq!(instance.get_focused_index(), 4);
+slint_testing::send_keyboard_string_sequence(&instance, "\t");
+assert_eq!(instance.get_focused_index(), 5);
+slint_testing::send_keyboard_string_sequence(&instance, "\t");
+assert_eq!(instance.get_focused_index(), 0);
```
Comment (LeonMatthes) [inline tests/cases/issues/issue_10321_tab_focus_flickable.slint]: This should probably be a loop as well.
```suggestion
for i in 0..=5 {
    slint_testing::send_keyboard_string_sequence(&instance, "\t");
    assert_eq!(instance.get_focused_index(), i);
}
slint_testing::send_keyboard_string_sequence(&instance, "\t");
assert_eq!(instance.get_focused_index(), 0);
```
